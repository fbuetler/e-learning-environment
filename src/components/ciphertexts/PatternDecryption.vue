<template>
  <div>
    <slot name="animation" :animationSteps="animationSteps" />
    <Difficulty
      :selected="currentDifficultyLevel"
      :difficultyLevels="difficultyLevels"
      @difficulty-selected="currentDifficultyLevel = $event"
    />
    <div class="flex-item flex-col">
      <div>Entschlüssle den Text mit Hilfe des Musters!</div>
      <div class="flex-item flex-row flex-center">
        <div class="flex-item flex-row flex-center equal-space">
          <div>Text:</div>
          <div id="text" class="canvas-container card big-text">
            {{ encryptedText }}
          </div>
        </div>
        <div class="flex-item flex-row flex-center equal-space">
          <div>Lösung:</div>
          <input
            id="answer-input"
            v-model="decryptedText"
            class="card big-text"
            size="5"
            type="text"
          />
        </div>
      </div>
      <div>
        <canvas id="pattern-canvas" class="card" width="500" height="150"
          >Das Muster ist nicht verfügbar</canvas
        >
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Mixins } from "vue-property-decorator";
import GameMixin, { GameInterface } from "@/components/GameMixins.vue";
import Difficulty from "@/components/Difficulty.vue";
import {
  CreatePattern,
  LoadRandomElement,
  PatternCanvas,
  Swap,
} from "@/components/ciphertexts/Ciphertext";

@Component<PatternDecryption>({
  components: {
    Difficulty,
  },
})
export default class PatternDecryption extends Mixins(GameMixin)
  implements GameInterface {
  dataKey = "nouns";

  originalText: string[] = null;
  decryptedText: string = null;

  patternPerLevel: Map<number, [number, number][]> = null;
  animationSteps: Array<string> = null;

  currentDifficultyLevel = 1;
  difficultyLevels = 2;

  mounted() {
    this.draw();
  }

  updated() {
    this.draw();
  }

  isStarted(): boolean {
    return this.originalText === null;
  }

  start() {
    this.originalText = LoadRandomElement(this.dataKey)
      .split("")
      .map((letter) => letter.toUpperCase());
    this.decryptedText = null;

    this.patternPerLevel = new Map<number, [number, number][]>();
    for (let level = 1; level <= this.difficultyLevels; level++) {
      this.patternPerLevel.set(level, CreatePattern(this.originalText, level));
    }
    this.animationSteps = this.getAnimationSteps();
  }

  isCorrect(): boolean {
    if (this.decryptedText === null) {
      return false;
    }
    return this.originalText.join("") === this.decryptedText.toUpperCase();
  }

  draw() {
    new PatternCanvas(
      document.getElementById(`pattern-canvas`) as HTMLCanvasElement
    ).draw(
      this.originalText.length,
      this.patternPerLevel.get(this.currentDifficultyLevel)
    );
  }

  getAnimationSteps(): Array<string> {
    const correctAnswer = this.originalText;
    let pattern: [number, number][];
    do {
      pattern = CreatePattern(correctAnswer, this.currentDifficultyLevel);
    } while (
      JSON.stringify(pattern) ===
      JSON.stringify(this.patternPerLevel.get(this.currentDifficultyLevel))
    );

    const wrongAnswer = Swap(correctAnswer.slice(), pattern);
    return [
      `answer-input:${wrongAnswer.join("")}`,
      "button-menu-check",
      `answer-input:${correctAnswer.join("")}`,
      "button-menu-check",
      "button-menu-next",
    ];
  }

  get encryptedText(): string {
    return Swap(
      this.originalText.slice(),
      this.patternPerLevel.get(this.currentDifficultyLevel)
    ).join("");
  }
}
</script>

<style scoped></style>
