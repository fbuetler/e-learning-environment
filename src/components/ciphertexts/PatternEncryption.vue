<template>
  <div>
    <slot name="animation" :animationSteps="animationSteps" />
    <Difficulty
      :selected="currentDifficultyLevel"
      :difficultyLevels="difficultyLevels"
      @difficulty-selected="currentDifficultyLevel = $event"
    />
    <div class="flex-item flex-col">
      <div>Verschlüssle den Text mit Hilfe des Musters!</div>
      <div class="flex-item flex-row flex-center">
        <div class="flex-item flex-row flex-center equal-space">
          <div>Text:</div>
          <div id="text" class="canvas-container card big-text">
            {{ decryptedText }}
          </div>
        </div>
        <div class="flex-item flex-row flex-center equal-space">
          <div>Lösung:</div>
          <input
            id="answer-input"
            v-model="encryptedText"
            class="card big-text"
            size="5"
            type="text"
          />
        </div>
      </div>
      <div>
        <canvas id="pattern-canvas" class="card" width="500" height="150"
          >Das Muster ist nicht verfügbar</canvas
        >
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Mixins } from "vue-property-decorator";
import GameMixin, { GameInterface } from "@/components/GameMixins.vue";
import Difficulty from "@/components/Difficulty.vue";
import {
  CreatePattern,
  LoadRandomElement,
  PatternCanvas,
  Swap,
} from "@/components/ciphertexts/Ciphertext";

@Component<PatternEncryption>({
  components: {
    Difficulty,
  },
})
export default class PatternEncryption extends Mixins(GameMixin)
  implements GameInterface {
  dataKey = "nouns";

  originalText: string[] = null;
  encryptedText: string = null;

  patternPerLevel: Map<number, [number, number][]> = null;
  animationSteps: Array<string> = null;

  currentDifficultyLevel = 1;
  difficultyLevels = 2;

  mounted() {
    this.draw();
  }

  updated() {
    this.draw();
  }

  isStarted(): boolean {
    return this.originalText === null;
  }

  start() {
    this.originalText = LoadRandomElement(this.dataKey)
      .split("")
      .map((letter) => letter.toUpperCase());
    this.encryptedText = null;

    this.patternPerLevel = new Map<number, [number, number][]>();
    for (let level = 1; level <= this.difficultyLevels; level++) {
      this.patternPerLevel.set(level, CreatePattern(this.originalText, level));
    }
    this.animationSteps = this.getAnimationSteps();
  }

  isCorrect(): boolean {
    if (this.encryptedText === null) {
      return false;
    }
    return this.encryptText === this.encryptedText.toUpperCase();
  }

  draw() {
    new PatternCanvas(
      document.getElementById(`pattern-canvas`) as HTMLCanvasElement
    ).draw(
      this.originalText.length,
      this.patternPerLevel.get(this.currentDifficultyLevel)
    );
  }

  getAnimationSteps(): Array<string> {
    const correctAnswer = this.encryptText;
    let pattern: [number, number][];
    do {
      pattern = CreatePattern(
        correctAnswer.split(""),
        this.currentDifficultyLevel
      );
    } while (
      JSON.stringify(pattern) ===
      JSON.stringify(this.patternPerLevel.get(this.currentDifficultyLevel))
    );

    const wrongAnswer = Swap(correctAnswer.split("").slice(), pattern).join("");
    return [
      `answer-input:${wrongAnswer}`,
      "button-menu-check",
      `answer-input:${correctAnswer}`,
      "button-menu-check",
      "button-menu-next",
    ];
  }

  get encryptText(): string {
    return Swap(
      this.originalText.slice(),
      this.patternPerLevel.get(this.currentDifficultyLevel)
    ).join("");
  }

  get decryptedText(): string {
    return this.originalText.join("");
  }
}
</script>

<style scoped></style>
