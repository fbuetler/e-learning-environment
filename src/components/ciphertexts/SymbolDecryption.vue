<template>
  <div>
    <slot name="animation" :animationSteps="animationSteps" />
    <Difficulty
      :selected="currentDifficultyLevel"
      :difficultyLevels="difficultyLevels"
      @difficulty-selected="
        currentDifficultyLevel = $event;
        start();
      "
    />
    <div class="flex-item flex-col">
      <div>Entschlüssle den Text mit Hilfe der Tabelle!</div>
      <div class="flex-item flex-row flex-center">
        <div class="flex-item flex-row flex-center equal-space">
          Text:
          <div id="text" class="canvas-container card">
            <canvas
              v-for="(part, index) in original"
              :id="'encrypted-' + index"
              :key="index"
              width="50"
              height="50"
              >{{ part }}</canvas
            >
          </div>
        </div>
        <div class="flex-flex equal-space">
          Lösung:
          <input
            id="answer-input"
            v-model="decrypted"
            class="card big-text"
            size="5"
            type="text"
          />
        </div>
      </div>
      <SymbolTable :table="table" :type="type" />
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Mixins } from "vue-property-decorator";
import GameMixin, { GameInterface } from "@/components/GameMixins.vue";
import Difficulty from "@/components/Difficulty.vue";
import SymbolTable from "@/components/ciphertexts/SymbolTable.vue";
import {
  SymbolConfig,
  LoadRandomNumber,
  LoadRandomElement,
  NumberTable,
  NumberLookup,
  LetterTable,
  LetterLookup,
  GetNewCanvas,
  Type,
} from "@/components/ciphertexts/Ciphertext";

@Component<SymbolDecryption>({
  components: {
    Difficulty,
    SymbolTable,
  },
})
export default class SymbolDecryption extends Mixins(GameMixin)
  implements GameInterface {
  dataKey = "nouns";

  originalNumbers: string[] = null;
  originalLetters: string[] = null;
  decrypted: string = null;
  animationSteps: Array<string> = null;

  currentDifficultyLevel = 1;
  difficultyLevels = 2;

  mounted() {
    this.drawShapes();
  }

  updated() {
    this.drawShapes();
  }

  isInitialized(): boolean {
    return this.originalNumbers !== null && this.originalLetters !== null;
  }

  start() {
    this.originalNumbers = String(LoadRandomNumber()).split("");
    this.decrypted = null;

    this.originalLetters = LoadRandomElement(this.dataKey)
      .split("")
      .map((letter) => letter.toUpperCase());
    this.animationSteps = this.getAnimationSteps();
  }

  isCorrect(): boolean {
    if (this.currentDifficultyLevel === 1) {
      return this.originalNumbers.join("") === this.decrypted;
    } else {
      return this.originalLetters.join("") === this.decrypted.toUpperCase();
    }
  }

  drawShapes() {
    this.original.forEach((part, index) => {
      const cvText = GetNewCanvas(
        this.type,
        document.getElementById(`encrypted-${index}`) as HTMLCanvasElement
      );
      cvText.draw(this.lookup.get(part));
    });
  }

  getAnimationSteps(): Array<string> {
    let wrongAnswer: string;
    if (this.currentDifficultyLevel === 1) {
      wrongAnswer = String(LoadRandomNumber());
    } else {
      wrongAnswer = Math.random()
        .toString(36)
        .replace(/[^a-z]+/g, "")
        .substr(0, this.originalLetters.length)
        .toUpperCase();
    }
    return [
      `answer-input:${wrongAnswer}`,
      "button-menu-check",
      `answer-input:${
        this.currentDifficultyLevel === 1
          ? this.originalNumbers.join("")
          : this.originalLetters.join("")
      }`,
      "button-menu-check",
      "button-menu-next",
    ];
  }

  get original(): string[] {
    return this.currentDifficultyLevel === 1
      ? this.originalNumbers
      : this.originalLetters;
  }

  get table(): SymbolConfig[][] {
    return this.currentDifficultyLevel === 1 ? NumberTable : LetterTable;
  }

  get lookup(): Map<string, SymbolConfig> {
    return this.currentDifficultyLevel === 1 ? NumberLookup : LetterLookup;
  }

  get type(): Type {
    return this.currentDifficultyLevel === 1 ? Type.NUMBER : Type.LETTER;
  }
}
</script>

<style scoped>
.flex-item {
  padding: 1em;
}
.canvas-container {
  margin: 1em;
}
.canvas-container > canvas {
  margin: 0.5em;
}
</style>
