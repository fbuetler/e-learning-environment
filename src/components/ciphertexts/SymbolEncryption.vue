<template>
  <div @dragend.prevent="selected = null">
    <slot name="animation" :animationSteps="animationSteps" />
    <div class="flex-item flex-col">
      <div>Verschl√ºssle den Text mit Hilfe der Tabelle!</div>
      <div class="flex-item flex-row flex-center">
        <div class="flex-item flex-row flex-center">
          <div>Text:</div>
          <div id="text" class="canvas-container card big-text">
            {{ originalText }}
          </div>
        </div>
        <div
          id="dropzone"
          class="flex-item flex-center flex-row flex-flex dropzone"
          @click="addSymbol()"
          @dragover.prevent
          @dragend.prevent
          @drop.stop.prevent="addSymbol()"
        >
          <div v-if="encryptedText.length === 0">
            Platziere hier die Symbole
          </div>
          <div id="encrypted-container" class="canvas-container"></div>
        </div>
        <Undo @undo-operation="undo()" />
      </div>
      <SymbolTable
        :table="table"
        :type="type"
        :selected="selected"
        @symbol-selected="
          (rowIndex, shapeIndex) => (selected = [rowIndex, shapeIndex])
        "
      />
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Mixins } from "vue-property-decorator";
import GameMixin, { GameInterface } from "@/components/GameMixins.vue";
import Undo from "@/components/Undo.vue";
import SymbolTable from "@/components/ciphertexts/SymbolTable.vue";
import {
  SymbolConfig,
  LoadRandomElement,
  LetterTable,
  LetterLookup,
  GetNewCanvas,
  Type,
  Shape,
} from "@/components/ciphertexts/Ciphertext";

@Component<SymbolEncryption>({
  components: {
    SymbolTable,
    Undo,
  },
})
export default class SymbolEncryption extends Mixins(GameMixin)
  implements GameInterface {
  dataKey = "nouns";

  originalText: string = null;
  encryptedText: string[] = null;

  selected: [number, number] = null;
  animationSteps: Array<string> = null;

  updated() {
    this.drawShapes();
  }

  isStarted(): boolean {
    return this.originalText === null;
  }

  restart() {
    this.originalText = LoadRandomElement(this.dataKey).toUpperCase();
    this.encryptedText = new Array<string>();
    this.deleteAllChildren(document.getElementById("encrypted-container"));
    this.animationSteps = this.getAnimationSteps();
  }

  isCorrect(): boolean {
    return this.originalText === this.encryptedText.join("");
  }

  undo() {
    this.encryptedText = new Array<string>();
    this.deleteAllChildren(document.getElementById("encrypted-container"));
  }

  addSymbol() {
    if (this.selected === null) {
      return;
    }
    this.appendCanvas(document.getElementById("encrypted-container"));
    this.encryptedText.push(
      this.getSymboltext(this.selected[0], this.selected[1])
    );
    this.selected = null;
  }

  appendCanvas(container: HTMLElement) {
    if (container === null) {
      return;
    }
    const id = container.childElementCount;
    const canvas = document.createElement("canvas") as HTMLCanvasElement;
    canvas.setAttribute("id", `encrypted-${id}`);
    canvas.setAttribute("width", "50");
    canvas.setAttribute("height", "50");
    canvas.innerText = "encrypted symbol";
    container.appendChild(canvas);
  }

  deleteAllChildren(container: HTMLElement) {
    if (container === null) {
      return;
    }
    container.innerHTML = "";
  }

  drawShapes() {
    this.encryptedText.forEach((part, index) => {
      const cvText = GetNewCanvas(
        this.type,
        document.getElementById(`encrypted-${index}`) as HTMLCanvasElement
      );
      cvText.draw(this.lookup.get(part));
    });
  }

  getAnimationSteps(): Array<string> {
    const correctText = this.originalText;
    let wrongText: string;
    do {
      wrongText = Math.random()
        .toString(36)
        .replace(/[^a-z]+/g, "")
        .substr(0, correctText.length)
        .toUpperCase();
    } while (wrongText === correctText);

    const mapToEncrypted = (word: string): string[] => {
      return word.split("").flatMap((el) => {
        const [rowIndex, shapeIndex] = this.findSymbol(el);
        return [`shape-${rowIndex}-${shapeIndex}`, "dropzone"];
      });
    };

    return mapToEncrypted(wrongText)
      .concat(["button-menu-check", "undo"])
      .concat(mapToEncrypted(correctText))
      .concat(["button-menu-check", "button-menu-next"]);
  }

  findSymbol(symbol: string): [number, number] {
    for (let i = 0; i < this.table.length; i++) {
      for (let j = 0; j < this.table[i].length; j++) {
        if (symbol === this.getSymboltext(i, j)) {
          return [i, j];
        }
      }
    }
    // should never be reached
    return [0, 0];
  }

  getSymboltext(rowIndex: number, shapeIndex: number): string {
    const symbol = this.table[rowIndex][shapeIndex].find(
      (shape) => shape[0] === Shape.TEXT
    );
    return symbol !== undefined ? (symbol[1].get("text") as string) : "";
  }

  get original(): string[] {
    return this.originalText.split("");
  }

  get table(): SymbolConfig[][] {
    return LetterTable;
  }

  get lookup(): Map<string, SymbolConfig> {
    return LetterLookup;
  }

  get type(): Type {
    return Type.LETTER;
  }
}
</script>

<style scoped>
.dropzone {
  min-height: 5em;
}
</style>
