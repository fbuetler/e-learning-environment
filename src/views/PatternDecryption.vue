<template>
  <Game :type="type">
    <h2 slot="title">{{ title }}</h2>
    <p slot="description">
      Biberin Anita findet auf einem Stück Rinde einen Geheimtext und Biberin
      Ada zeigt ihr ein anderes Stück Rinde und meint darauf ist "das Geheimnis
      der Geheimschrift". Der Text wurde chiffriert, indem die Buchstaben nach
      einem bestimmten Muster ausgetauscht wurden. Schaffst du es, den
      Geheimtext zu lesen?
    </p>
    <video loop controls muted slot="video">
      <source :src="require('@/assets/tutorials/' + video)" />
      Your browser does not support the video tag.
    </video>
  </Game>
</template>

<script lang="ts">
import Vue from "vue";
import { Component } from "vue-property-decorator";
import Game, { GameType } from "@/components/Game.vue";
import views from "@/views/Views";

@Component<PatternDecryption>({
  components: {
    Game,
  },
})
export default class PatternDecryption extends Vue {
  type: GameType = GameType.PATTERNDECRYPTION;
  title = views.find((el) => el.component === GameType.PATTERNDECRYPTION).title;
  video = "patterndecryption.mp4";
}
</script>
