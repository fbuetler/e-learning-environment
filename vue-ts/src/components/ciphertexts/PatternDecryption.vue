<template>
  <div>
    <canvas id="test-canvas" width="500" height="150">test</canvas>
  </div>
</template>

<script lang="ts">
import Vue from "vue";
import { Component } from "vue-property-decorator";
import { PatternCanvas } from "./Ciphertext";

@Component<PatternDecryption>({})
export default class PatternDecryption extends Vue {
  mounted() {
    this.draw();
  }

  updated() {
    this.draw();
  }

  draw() {
    const patternCanvas = new PatternCanvas(
      document.getElementById(`test-canvas`) as HTMLCanvasElement
    );

    patternCanvas.draw(15, [
      [0, 14],
      [1, 7],
      [3, 5],
      [2, 6],
      [4, 13],
      [8, 12],
      [9, 11],
    ]);
  }
}
</script>

<style scoped></style>
