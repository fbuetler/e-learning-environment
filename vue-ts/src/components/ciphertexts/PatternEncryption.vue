<template>
  <div>
    <Difficulty
      :selected="currentDifficultyLevel"
      :difficultyLevels="difficultyLevels"
      @difficulty-selected="changeDifficulty($event)"
    />
    <div class="flex-item flex-col">
      <div>Verschlüssle den Text mit Hilfe des Musters!</div>
      <div class="flex-item flex-row flex-center">
        <div class="flex-item flex-row flex-center equal-space">
          <div>Text:</div>
          <div class="canvas-container card text">
            {{ decryptedText }}
          </div>
        </div>
        <div class="flex-item flex-row flex-center equal-space">
          <div>Lösung:</div>
          <input class="card text" v-model="encryptedText" type="text" />
        </div>
      </div>
      <div>
        <canvas class="card" id="pattern-canvas" width="500" height="150"
          >Das Muster ist nicht verfügbar</canvas
        >
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Mixins } from "vue-property-decorator";
import GameMixin, { GameInterface } from "../Game";
import Difficulty from "../Difficulty.vue";
import { CreatePattern, LoadRandomElement, PatternCanvas } from "./Ciphertext";

@Component<PatternEncryption>({
  components: {
    Difficulty,
  },
})
export default class PatternEncryption extends Mixins(GameMixin)
  implements GameInterface {
  dataKey = "nouns";

  originalText: string[] = null;
  encryptedText: string = null;

  patterns: Map<number, [number, number][]> = null;

  currentDifficultyLevel: number = null;
  difficultyLevels = 2;

  mounted() {
    this.draw();
  }

  updated() {
    this.draw();
  }

  isStarted(): boolean {
    return this.originalText === null;
  }

  restartGame() {
    if (this.currentDifficultyLevel === null) {
      this.currentDifficultyLevel = 1;
    }
    this.originalText = LoadRandomElement(this.dataKey)
      .split("")
      .map((letter) => letter.toUpperCase());
    this.encryptedText = null

    this.patterns = new Map<number, [number, number][]>();
    for (let level = 1; level <= this.difficultyLevels; level++) {
      this.patterns.set(level, CreatePattern(this.originalText, level));
    }
  }

  isCorrect(): boolean {
    if (this.encryptedText === null) {
      return false;
    }
    const text = JSON.parse(JSON.stringify(this.originalText)) as string[];
    this.patterns.get(this.currentDifficultyLevel).forEach(([left, right]) => {
      const tmp = text[left];
      text[left] = text[right];
      text[right] = tmp;
    });
    return text.join("") === this.encryptedText.toUpperCase();
  }

  draw() {
    new PatternCanvas(
      document.getElementById(`pattern-canvas`) as HTMLCanvasElement
    ).draw(
      this.originalText.length,
      this.patterns.get(this.currentDifficultyLevel)
    );
  }

  changeDifficulty(level: number) {
    this.currentDifficultyLevel = level;
  }

  get decryptedText(): string {
    return this.originalText.join("");
  }
}
</script>

<style scoped>
.text {
  font-size: 2em;
}
</style>
