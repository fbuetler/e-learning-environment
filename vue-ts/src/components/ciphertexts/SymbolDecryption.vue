<template>
  <div>
    <Difficulty
      :selected="currentDifficultyLevel"
      :difficultyLevels="difficultyLevels"
      @difficulty-selected="changeDifficulty($event)"
    />
    <div class="flex-item flex-col">
      <div>Entschlüssle den Text mit Hilfe der Tabelle</div>
      <div class="flex-item flex-row flex-center">
        <div class="flex-item flex-row flex-center equal-space">
          Text:
          <div class="canvas-container">
            <canvas
              v-for="(part, index) in text"
              :key="index"
              :id="'encrypted-text-' + index"
              >Text</canvas
            >
          </div>
        </div>
        <div class="flex-flex equal-space">
          Lösung:
          <input class="card" v-model="decryptedText" type="text" />
        </div>
      </div>
      <div class="flex-item flex-wrap">
        <div
          class="flex flex-center"
          v-for="(shape, index) in table"
          :key="index"
        >
          <canvas :id="'shape-' + index">Shape</canvas>
        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Mixins } from "vue-property-decorator";
import GameMixin, { GameInterface } from "../Game";
import Difficulty from "../Difficulty.vue";
import {
  NumberCanvas,
  LetterCanvas,
  Shape,
  LoadRandomNumber,
  LoadRandomElement,
  NumberTable,
  NumberLookup,
  LetterTable,
  LetterLookup,
} from "./Ciphertext";

@Component<SymbolDecryption>({
  components: {
    Difficulty,
  },
})
export default class SymbolDecryption extends Mixins(GameMixin)
  implements GameInterface {
  dataKey = "nouns";

  originalText: string = null;
  decryptedText: string = null;

  lookupTable: Map<string, [Shape, Map<string, number>][]> = null;
  table: [Shape, Map<string, number>][][] = null;

  currentDifficultyLevel: number = null;
  difficultyLevels = 2;

  mounted() {
    this.drawShapes();
  }

  updated() {
    this.drawShapes();
  }

  isStarted(): boolean {
    return this.originalText === null;
  }

  restartGame() {
    if (this.currentDifficultyLevel === null) {
      this.currentDifficultyLevel = 1;
    }
    if (this.currentDifficultyLevel === 1) {
      this.originalText = String(LoadRandomNumber());
      this.lookupTable = NumberLookup;
      this.table = NumberTable;
    } else {
      this.originalText = LoadRandomElement(this.dataKey);
      this.lookupTable = LetterLookup;
      this.table = LetterTable;
    }
  }

  isCorrect(): boolean {
    return this.originalText == this.decryptedText;
  }

  drawShapes() {
    this.text.forEach((number, index) => {
      const cvText = new NumberCanvas(
        document.getElementById(`encrypted-text-${index}`) as HTMLCanvasElement,
        100,
        100
      );
      cvText.draw(this.lookupTable.get(number));
    });
    this.table.forEach((shapes, index) => {
      const cvShape = new NumberCanvas(
        document.getElementById(`shape-${index}`) as HTMLCanvasElement,
        100,
        100
      );
      cvShape.draw(shapes);
    });
  }

  changeDifficulty(level: number) {
    this.currentDifficultyLevel = level;
  }

  get text(): string[] {
    return this.originalText.toString().split("");
  }
}
</script>

<style scoped>
.flex {
  flex: 0 0 15%; /* ensures having 5 cols per row */
  padding: 1em;
  border-right: 4px solid black;
  border-bottom: 4px solid black;
  color: orange;
}
.flex-item {
  padding: 1em;
}
.equal-space {
  flex: 1 1 0px;
}
.canvas-container {
  margin: 1em;
}
</style>
