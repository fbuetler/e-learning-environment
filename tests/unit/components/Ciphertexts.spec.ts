import { shallowMount, config } from "@vue/test-utils";
import PatternEncryption from "@/components/ciphertexts/PatternEncryption.vue";
import PatternDecryption from "@/components/ciphertexts/PatternDecryption.vue";
import SymbolEncryption from "@/components/ciphertexts/SymbolEncryption.vue";
import SymbolDecryption from "@/components/ciphertexts/SymbolDecryption.vue";

config.showDeprecationWarnings = false;

const mockText = "Test";
const mockPatternOne = [[1, 2]] as [number, number][];
const mockPatternTwo = [
  [0, 1],
  [2, 3],
] as [number, number][];
const mockEncryptedTextOne = "TSET";
const mockEncryptedTextTwo = "ETTS";

const mockDraw = jest.fn();
const mockGetAnimationSteps = jest.fn();
jest.mock("@/components/ciphertexts/Ciphertext", () => ({
  LoadRandomElement(key: string): string {
    return mockText;
  },
  CreatePattern(text: string[], swapAmount: number): Array<[number, number]> {
    if (swapAmount === 1) {
      return mockPatternOne;
    } else if (swapAmount === 2) {
      return mockPatternTwo;
    } else {
      return [];
    }
  },
  Swap(str: string[], pattern: [number, number][]): string[] {
    pattern.forEach(([left, right]) => {
      const tmp = str[left];
      str[left] = str[right];
      str[right] = tmp;
    });
    return str;
  },
}));

describe("PatternEncryption.vue", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(PatternEncryption, {
      methods: { draw: mockDraw, getAnimationSteps: mockGetAnimationSteps },
    });
  });

  it("is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("initial conditions hold", () => {
    expect(wrapper.vm["decryptedText"]).toBe(mockText.toUpperCase());
    expect(wrapper.vm["encryptedText"]).toBeNull();
  });

  it("next task restores initial conditions", async () => {
    await wrapper.setData({ encryptedText: "ABC" });
    wrapper.vm.restartGame();
    expect(wrapper.vm["encryptedText"]).toBeNull();
  });

  it("correct answer is accepted", async () => {
    await wrapper.setData({ encryptedText: mockEncryptedTextOne });
    expect(wrapper.vm.isCorrect()).toBeTruthy();
  });

  it("incorrect answer is rejected", async () => {
    await wrapper.setData({ encryptedText: "ABC" });
    expect(wrapper.vm.isCorrect()).toBeFalsy();
  });
});

describe("PatternDecryption.vue", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(PatternDecryption, {
      methods: { draw: mockDraw, getAnimationSteps: mockGetAnimationSteps },
    });
  });

  it("is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("initial conditions hold", () => {
    expect(wrapper.vm["decryptedText"]).toBeNull();
    expect(wrapper.vm["encryptedText"]).toBe(mockEncryptedTextOne);
  });

  it("next task restores initial conditions", async () => {
    await wrapper.setData({ decryptedText: "ABC " });
    wrapper.vm.restartGame();
    expect(wrapper.vm["decryptedText"]).toBeNull();
    expect(wrapper.vm["encryptedText"]).toBe(mockEncryptedTextOne);
  });

  it("correct answer is accepted", async () => {
    await wrapper.setData({ decryptedText: mockText });
    expect(wrapper.vm.isCorrect()).toBeTruthy();
  });

  it("incorrect answer is rejected", async () => {
    await wrapper.setData({ decryptedText: "ABC " });
    expect(wrapper.vm.isCorrect()).toBeFalsy();
  });
});

describe("SymbolEncryption.vue", () => {
  // importing enum and classes is broken in ts-jest
});

describe("SymbolDecryption.vue", () => {
  // importing enum and classes is broken in ts-jest
});
