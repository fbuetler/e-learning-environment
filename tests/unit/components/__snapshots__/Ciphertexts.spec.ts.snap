// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PatternDecryption.vue renders correctly 1`] = `
<div>
   
  <difficulty-stub
    difficultylevels="2"
    selected="1"
  />
   
  <div
    class="flex-item flex-col"
  >
    <div>
      Entschlüssle den Text mit Hilfe des Musters!
    </div>
     
    <div
      class="flex-item flex-row flex-center"
    >
      <div
        class="flex-item flex-row flex-center equal-space"
      >
        <div>
          Text:
        </div>
         
        <div
          class="canvas-container card big-text"
          id="text"
        >
          
          TSET
        
        </div>
      </div>
       
      <div
        class="flex-item flex-row flex-center equal-space"
      >
        <div>
          Lösung:
        </div>
         
        <input
          class="card big-text"
          id="answer-input"
          size="5"
          type="text"
        />
      </div>
    </div>
     
    <div>
      <canvas
        class="card"
        height="150"
        id="pattern-canvas"
        width="500"
      >
        Das Muster ist nicht verfügbar
      </canvas>
    </div>
  </div>
</div>
`;

exports[`PatternEncryption.vue renders correctly 1`] = `
<div>
   
  <difficulty-stub
    difficultylevels="2"
    selected="1"
  />
   
  <div
    class="flex-item flex-col"
  >
    <div>
      Verschlüssle den Text mit Hilfe des Musters!
    </div>
     
    <div
      class="flex-item flex-row flex-center"
    >
      <div
        class="flex-item flex-row flex-center equal-space"
      >
        <div>
          Text:
        </div>
         
        <div
          class="canvas-container card big-text"
          id="text"
        >
          
          TEST
        
        </div>
      </div>
       
      <div
        class="flex-item flex-row flex-center equal-space"
      >
        <div>
          Lösung:
        </div>
         
        <input
          class="card big-text"
          id="answer-input"
          size="5"
          type="text"
        />
      </div>
    </div>
     
    <div>
      <canvas
        class="card"
        height="150"
        id="pattern-canvas"
        width="500"
      >
        Das Muster ist nicht verfügbar
      </canvas>
    </div>
  </div>
</div>
`;
